src/Dialect/ONNX/ONNXOps.cpp

```diff
@@ -2372,6 +2417,51 @@ LogicalResult ONNXConcatOp::inferShapes() {
         j == axisIndex ? cummulativeAxisSize : commonShape[j]);
   getResult().setType(
       RankedTensorType::get(outputDims, commonType.getElementType()));
+
+  // Checking to see if shape folding is possible
+  bool canShapeFold = false;
+  int64_t size = 0;
+  for (int i = 0; i < inputNum; ++i) {
+    if (auto constantOp = getONNXConstantOp(getOperand(i))) {
+      if (DenseElementsAttr shapeAttributes =
+              constantOp.valueAttr().dyn_cast_or_null<DenseElementsAttr>()) {
+        size += shapeAttributes.size();
+      } else {
+        return emitError("Failed passing shape folding information");
+      }
+    } else if (auto valueAttribute = getShapeFoldingAttr(getOperand(i))) {
+      size += valueAttribute.size();
+    }
+  }
+  // TODO: support more than 1D Tensor
+  if (axisIndex == 0 && outputDims[0] == size)
+    canShapeFold = true;
+
+  // Only two ways we can get shape folding to work here
+  // 1. The shape information(data of the operand) is stored in shapeMap already
+  // 2. The operand is a contant Op
+  if (canShapeFold) {
+    SmallVector<int64_t, 4> outShape;
+    for (int i = 0; i < inputNum; ++i) {
+      if (auto constantOp = getONNXConstantOp(getOperand(i))) {
+        DenseElementsAttr shapeAttributes =
+            constantOp.valueAttr().dyn_cast_or_null<DenseElementsAttr>();
+        auto attrIt = shapeAttributes.getValues<IntegerAttr>().begin();
+        for (int j = 0; j != shapeAttributes.size(); ++j)
+          outShape.push_back((*attrIt++).cast<IntegerAttr>().getInt());
+      } else if (auto valueAttribute = getShapeFoldingAttr(getOperand(i))) {
+        auto valueIt = valueAttribute.getValues<IntegerAttr>().begin();
+        for (int i = 0; i != valueAttribute.size(); ++i)
+          outShape.push_back((*valueIt++).cast<IntegerAttr>().getInt());
+      }
+    }
+    // Only save shape folding attribute if there is no missing shape
+    // information
+    setAttr("shape_folding_value",
```

Comment:
Shall we refactor this to a `setShapeFoldingAttribute`?

Reply:
Yes. We can do that. Do you think it is appropriate to make it a member function of Op or do you think we should an Operation*

Reply:
We can have both `getShapeFoldingAttr` and `setShapeFoldingAttr` as methods of `ShapeInferencePass`.

Reply:
Right now this is no header file for ShapeInferencePass. So I put it in ONNXShapeHelper.hpp

Reply:
Sounds good.